streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -671,18 +688,35 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
   final def flatMap[R1 <: R, E1 >: E, B](f0: A => ZStream[R1, E1, B]): ZStream[R1, E1, B] =
     new ZStream[R1, E1, B] {
       def fold[R2 <: R1, E2 >: E1, B1 >: B, S]: Fold[R2, E2, B1, S] =
-        ZManaged.succeed { (s, cont, f) =>
-          self.fold[R2, E2, A, S].flatMap { foldOuter =>
-            foldOuter(s, cont, (s, a) => {
-              f0(a)
-                .fold[R2, E2, B1, S]
-                .flatMap { foldInner =>
-                  foldInner(s, cont, f)
+        foldDefault
+
+      override def process: ZManaged[R1, E1, ZStream.InputStream[R1, E1, B]] =
+        for {
+          ref <- Ref.make[Option[(InputStream[R1, E1, B], Exit[_, _] => ZIO[R1, Nothing, Any])]](None).toManaged_
+          as  <- self.process
+          _   <- ZManaged.finalizerExit(e => ref.get.flatMap(_.map(_._2).getOrElse((_: Exit[_, _]) => UIO.unit).apply(e)))
+          pullOuter = ZIO.uninterruptibleMask { restore =>
+            restore(as).flatMap { a =>
+              (for {
+                reservation <- f0(a).process.reserve
+                bs          <- restore(reservation.acquire)
+                _           <- ref.set(Some(bs -> reservation.release))
+              } yield ()).mapError(Some(_))
+            }
+          }
+          bs = {
+            def go: InputStream[R1, E1, B] = ref.get.flatMap {
+              case None => pullOuter *> go
+              case Some((isB, finalizer)) =>
+                isB.catchAll {
+                  case e @ Some(e1) => (finalizer(Exit.fail(e1)) *> ref.set(None)).uninterruptible *> ZIO.fail(e)
+                  case None         => (finalizer(Exit.succeed(())) *> ref.set(None)).uninterruptible *> pullOuter *> go
```

Comment:
Ensuring the `finalizer` is not interrupted is one thing, we also really need to ensure the `finalizer` will be run in all cases, i.e. that `go` will not be interrupted right before it even begins execution of the finalizer.

Reply:
Ouch, good point. Will use `uninterruptibleMask` here too.

Reply:
@jdegoes Oh, I actually handled this. If `go` gets interrupted, the `ZManaged` will execute the stored finalizer (line 697).